cirq-google/cirq_google/serialization/arg_func_langs.py

```diff
@@ -176,6 +187,75 @@ def arg_to_proto(
+    match value.dtype:
+        case np.float64:
+            ndarrays.to_float64_array(value, out=ndarray_msg.float64_array)
+        case np.float32:
+            ndarrays.to_float32_array(value, out=ndarray_msg.float32_array)
+        case np.float16:
+            ndarrays.to_float16_array(value, out=ndarray_msg.float16_array)
+        case np.int64:
+            ndarrays.to_int64_array(value, out=ndarray_msg.int64_array)
+        case np.int32:
+            ndarrays.to_int32_array(value, out=ndarray_msg.int32_array)
+        case np.int16:
+            ndarrays.to_int16_array(value, out=ndarray_msg.int16_array)
+        case np.int8:
+            ndarrays.to_int8_array(value, out=ndarray_msg.int8_array)
+        case np.uint8:
+            ndarrays.to_uint8_array(value, out=ndarray_msg.uint8_array)
+        case np.complex128:
+            ndarrays.to_complex128_array(value, out=ndarray_msg.complex128_array)
+        case np.complex64:
+            ndarrays.to_complex64_array(value, out=ndarray_msg.complex64_array)
+        case np.bool_:
+            ndarrays.to_bitarray(value, out=ndarray_msg.bit_array)
+
+
+def _ndarray_from_proto(msg: v2.program_pb2.ArgValue):
+    ndarray_msg = msg.ndarray_value
+    match ndarray_msg.WhichOneof('arr'):
+        case 'float64_array':
+            return ndarrays.from_float64_array(ndarray_msg.float64_array)
+        case 'float32_array':
+            return ndarrays.from_float32_array(ndarray_msg.float32_array)
+        case 'float16_array':
+            return ndarrays.from_float16_array(ndarray_msg.float16_array)
+        case 'int64_array':
+            return ndarrays.from_int64_array(ndarray_msg.int64_array)
+        case 'int32_array':
+            return ndarrays.from_int32_array(ndarray_msg.int32_array)
+        case 'int16_array':
+            return ndarrays.from_int16_array(ndarray_msg.int16_array)
+        case 'int8_array':
+            return ndarrays.from_int8_array(ndarray_msg.int8_array)
+        case 'uint8_array':
+            return ndarrays.from_uint8_array(ndarray_msg.uint8_array)
+        case 'complex128_array':
+            return ndarrays.from_complex128_array(ndarray_msg.complex128_array)
+        case 'complex64_array':
+            return ndarrays.from_complex64_array(ndarray_msg.complex64_array)
+        case 'bit_array':
+            return ndarrays.from_bitarray(ndarray_msg.bit_array)
+
+
+def _tuple_to_proto(value: Union[list, tuple, set, frozenset], out: v2.program_pb2.Tuple):
+    """Converts a tuple of mixed values to Arg protos."""
+    if isinstance(value, list):
+        out.sequence_type = v2.program_pb2.Tuple.SequenceType.LIST
+    elif isinstance(value, tuple):
+        out.sequence_type = v2.program_pb2.Tuple.SequenceType.TUPLE
+    elif isinstance(value, set):
+        out.sequence_type = v2.program_pb2.Tuple.SequenceType.SET
+    elif isinstance(value, frozenset):
+        out.sequence_type = v2.program_pb2.Tuple.SequenceType.FROZENSET
```

Comment:
Nit - maybe we should reset sequence_type in case it was set in the `out` argument -
```suggestion
    elif isinstance(value, frozenset):
        out.sequence_type = v2.program_pb2.Tuple.SequenceType.FROZENSET
    else:
        out.sequence_type = v2.program_pb2.Tuple.SequenceType.UNSPECIFIED
```